File: 189.Rotate_Array/main.c
#include <stdio.h>

void reverse(int *nums, int start, int end)
{
     while (start < end)
     {
          int temp = nums[start];
          nums[start] = nums[end];
          nums[end] = temp;
          start++;
          end--;
     }
}

void rotate(int *nums, int numsSize, int k)
{
     k %= numsSize; // Handle cases where k >= numsSize

     reverse(nums, 0, numsSize - 1); // Step 1: Reverse the entire array
     reverse(nums, 0, k - 1);        // Step 2: Reverse the first k elements
     reverse(nums, k, numsSize - 1); // Step 3: Reverse the rest
}

int main()
{
     int nums[] = {1, 2, 3, 4, 5, 6, 7};
     int numsSize = sizeof(nums) / sizeof(nums[0]);
     rotate(nums, numsSize, 3);

     for (int i = 0; i < numsSize; i++)
     {
          printf("%d ", nums[i]);
     }
     // Output: 5 6 7 1 2 3 4
     return 0;
}

File: 189.Rotate_Array/main.go
package main

import "fmt"

func rotate(nums []int, k int) {
    n := len(nums)
    k %= n // Handle cases where k >= n

    reverse := func(start, end int) {
        for start < end {
            nums[start], nums[end] = nums[end], nums[start]
            start++
            end--
        }
    }

    reverse(0, n-1) // Step 1: Reverse the entire array
    reverse(0, k-1) // Step 2: Reverse the first k elements
    reverse(k, n-1) // Step 3: Reverse the rest
}

func main() {
    nums := []int{1, 2, 3, 4, 5, 6, 7}
    rotate(nums, 3)
    fmt.Println(nums) // Output: [5, 6, 7, 1, 2, 3, 4]
}
